packages/shared/getComponentNameFromType.js

```diff
@@ -90,7 +91,11 @@ export default function getComponentNameFromType(type: mixed): string | null {
       case REACT_FORWARD_REF_TYPE:
         return getWrappedName(type, type.render, 'ForwardRef');
       case REACT_MEMO_TYPE:
-        return getComponentNameFromType(type.type);
+        const outerName = (type: any).displayName || null;
+        if (outerName !== null) {
+          return outerName;
+        }
```

Comment:
I don't think we should have been doing this for `memo` and `forwardRef` because it's a side effect. We were doing it because of how it impacted component stacks but I'm not sure that's the right approach.

For instance:

```js
function MyComponent() {
  // ...
}

const MyMemoizedComponent = React.memo(MyComponent);
MyMemoizedComponent.displayName = 'Blah';

// Now if I render <MyComponent> it will also have a displayName of "blah"
// That doesn't seem good.
```

We could modify our component stack generation so support component stacks _without_ this side effect if it's worth it. I considered doing that change in this PR but I left it out initially.

Reply:
Maybe @sebma has an opinion on the stack frame thing.

Reply:
> Maybe @sebma has an opinion on the stack frame thing.

Did you mean to ping [sebmarkbage](http://github.com/sebmarkbage)?

Reply:
Oh, yeah! Auto-complete failed me there.

Reply:
Maybe a compromise would be to only set the display name if the inner component has no `name` as well (aka only for anonymous functions– since that strongly suggests e.g. `React.memo(props => {...})`)